stellargraph/data/explorer.py

```diff
@@ -47,6 +48,62 @@ def _default_if_none(value, default, name, ensure_not_none=True):
     return value
 
 
+def _ensure_int(value, name, min_value):
+    if not isinstance(value, int) or value < min_value:
+        raise ValueError(
+            f"{name}: expected an integer greater than or equal to {min_value}, found: {value}"
+        )
+
+
+class RandomWalk(ABC):
+    """
+    Abstract base class for Random Walk classes. A Random Walk class must implement a ``run`` method
+    which takes an iterable of node IDs and returns a list of walks. Each walk is a list of node IDs
+    that contains the starting node as its first element.
+    """
+
+    def __init__(self, graph, seed=None):
+        if not isinstance(graph, StellarGraph):
+            raise TypeError("Graph must be a StellarGraph or StellarDiGraph.")
+
+        self.graph = graph
+        self._random_state, self._np_random_state = random_state(seed)
+
+    def _get_random_state(self, seed):
+        """
+        Args:
+            seed: The optional seed value for a given run.
+
+        Returns:
+            The random state as determined by the seed.
+        """
+        if seed is None:
+            # Restore the random state
+            return self._random_state
+        # seed the random number generator
+        _ensure_int(seed, "seed", min_value=0)
+        rs, _ = random_state(seed)
+        return rs
+
+    @staticmethod
+    def _validate_walk_params(nodes, n, length):
+        if not is_real_iterable(nodes):
+            raise ValueError(f"nodes: expected an iterable, found: {nodes}")
+        if len(nodes) == 0:
+            warnings.warn(
+                "No root node IDs given. An empty list will be returned as a result.",
+                RuntimeWarning,
+                stacklevel=3,
+            )
+
+        _ensure_int(n, "n", min_value=1)
+        _ensure_int(length, "length", min_value=1)
+
+    @abstractmethod
+    def run(self, nodes, **kwargs):
```

Comment:
Also, theoretically the subclasses don't currently satisfy this signature because there's no `*args` here, or `*` on the subclasses. For instance, if `rw = UniformRandomWalk(...)`, then `rw.run(start_nodes, 10)` will work and pass 10, but that isn't allowed by the abstract signature here. We could either update this to be:

```suggestion
    def run(self, nodes, *args, **kwargs):
```

or update each of the run methods to make the non-`nodes` params keyword only, like 

```python
def run(self, nodes, *, n=None, length=None, seed=None):
```

Thoughts?

Reply:
> update each of the run methods to make the non-nodes params keyword only

I like this approach, I think we want to force anything other than `nodes` to be keyword only

Reply:
Sounds reasonable.

Reply:
Hmm tests still pass, but pycharm doesn't seem to like it...
![image](https://user-images.githubusercontent.com/33508488/78107602-83bb2e00-7441-11ea-9aad-0416bc1aebcc.png)
whereas it wasn't complaining before.

Reply:
Huh; I think that might be a PyCharm bug? Because it seems to be that the keyword-only one is the one that matches the behaviour of the abstract one:

```python
class X:
  def run_abstract(self, nodes, **kwargs):
    print(nodes, kwargs)
  def run_develop(self, nodes, n=None, length=None, seed=None):
    print(nodes, n, length, seed)
  def run_keyword_only(self, nodes, *, n=None, length=None, seed=None):
    print(nodes, n, length, seed)

x = X()

# positional
x.run_abstract([1], 2) # TypeError: run_abstract() takes 2 positional arguments but 3 were given
x.run_keyword_only([1], 2) # TypeError: run_keyword_only() takes 2 positional arguments but 3 were given
x.run_develop([1], 2) # [1] 2 None None

# keyword
x.run_abstract([1], n=2) # [1] {'n': 2}
x.run_keyword_only([1], n=2) # [1] 2 None None
x.run_develop([1], n=2) # [1] 2 None None
```

I wonder if there's a public bug tracker we can search/file against?

Reply:
There's a number of issues that seem somewhat related: https://youtrack.jetbrains.com/issues/PY?q=signature%20does%20not%20match%20base

I don't see one that specifically talks about using `*` though - I can adapt your example a bit and file an issue

Reply:
Hmm I upgraded pycharm and now the error seems to be gone actually... which I probably should've done before filing the issue 😅 Maybe the positional one should show an error though (which it still doesn't)